### PR TITLE
Update go version to 1.18.1

### PR DESCRIPTION
**Pull request by DanStough**
Updates to facilitate Go 1.18. Highlights on Changes:
* MOG updates which were released to deal with 1.18 changes
* Linter Update to Support 1.18 changes
* Misc Linter Fixes
* CI Runner updates
* Regenerated Test Certificates w/o using SHA1
* Deprecation Fixes

Deprecations:
* Verifying Certs w/ SHA1 signatures
* `x509.CertPool.Subjects()`
* `Strings.Title()`

### CODE REVIEW

File: connect/tls.go
Comment (DanStough) [inline connect/tls.go:125]: Sadly, this is the first time I've used `reflect`. Something about this seems like it needs a safety check but I didn't dig too deep to figure out where.

File: agent/consul/leader_connect_ca.go
Comment (rboyer) [inline agent/consul/leader_connect_ca.go:272]: It's funny that the only place we use this is to pretty print these strings:
```
const (
	ConsulCAProvider = "consul"  // -> "Consul"
	VaultCAProvider  = "vault"   // -> "Vault"
	AWSCAProvider    = "aws-pca" // -> "Aws-Pca" :-(
)
```

File: api/api_test.go
Comment (rboyer) [inline api/api_test.go:1110]:
```suggestion
	require.NoError(t, err)
```

File: api/api_test.go
Comment (rboyer) [inline api/api_test.go:1118]:
```suggestion
		require.NoError(t, err)
```

File: agent/agent_test.go
Comment (rboyer) [inline agent/agent_test.go:5865]:
```suggestion
	require.NoError(t, err)
```

File: agent/agent_test.go
Comment (rboyer) [inline agent/agent_test.go:5877]:
```suggestion
	require.NoError(t, err)
```

File: agent/agent_test.go
Comment (rboyer) [inline agent/agent_test.go:5885]:
```suggestion
		require.NoError(t, err)
```

File: tlsutil/config.go
Comment (rboyer) [inline tlsutil/config.go:465]:
```suggestion
// getSubjects uses the reflect package to test for rawSubject fields.
```

File: tlsutil/config.go
Comment (rboyer) [inline tlsutil/config.go:459]: I think I would actually prefer if we drop this conditional entirely so we don't need this reflect trickery in non-test code.

Comment (rboyer) [reply tlsutil/config.go:459]: cc: @i0rek  who might have more context on what this was for originally in #5597

File: .changelog/12808.txt
Comment (FFMMM) [inline .changelog/12808.txt:2]:
```suggestion
dependency: Upgrade to use Go 1.18.1
```

File: tlsutil/config.go
Comment (DanStough) [reply tlsutil/config.go:459]: If we're only worried that a cert was added and not that it had a valid name, this might work:
```golang
func newX509CertPool(groups ...[]string) (*x509.CertPool, error) {
	var haveCerts bool
	pool := x509.NewCertPool()
	for _, group := range groups {
		for _, pem := range group {
			if !pool.AppendCertsFromPEM([]byte(pem)) {
				return nil, fmt.Errorf("failed to parse PEM %s", pem)
			}
			if len(pem) > 0 {
				haveCerts = true
			}
		}
	}
	if !haveCerts {
		return nil, nil
	}
	return pool, nil
}
```

File: agent/consul/leader_connect_ca.go
Comment (FFMMM) [inline agent/consul/leader_connect_ca.go:17]: I'm not really sure if it's worth pulling this in 🤔 

For what we use it later it seems that it's redundant. I can see the full value here if Consul could support internationalization of strings but it doesn't yet. Unless you or the team feels strongly for this, I'd suggest we punt on this change 🙈 

Hopefully I didn't miss a critical use case for this 🤦🏼 🤞🏼

File: api/api_test.go
Comment (FFMMM) [inline api/api_test.go:1125]: this is 👨🏼‍🍳 👌🏼 , 💯 !

File: .github/workflows/build.yml
Comment (FFMMM) [inline .github/workflows/build.yml:176]: nothing actionable now but pointing out that this may cause merge conflicts in ENT 📚

File: agent/connect_ca_endpoint_test.go
Comment (FFMMM) [inline agent/connect_ca_endpoint_test.go:294]: do we not care/ want for this check anymore 🤔 ?

File: sdk/testutil/testlog.go
Comment (FFMMM) [inline sdk/testutil/testlog.go:10]: why ? 🙈

File: tlsutil/config.go
Comment (FFMMM) [inline tlsutil/config.go:467]: this feels like it could panic if there is no field `byName` right?

**Review by FFMMM — COMMENTED** [review]
This is awesome 💯 ! Thanks for working on this 🥇 ; and congrats on your first PR 🎉 !!

I left some questions and nits; It looks like you're working with RB so I'll defer to him for final approval.

Things LGTM overall!!

File: agent/consul/leader_connect_ca.go
Comment (DanStough) [reply agent/consul/leader_connect_ca.go:17]: Yeah, I think I was just trying to leave things as-is as much as possible. Per the Slack thread with @rboyer, I think we can just remove the use case.

File: api/api_test.go
Comment (DanStough) [reply api/api_test.go:1125]: Totally ~stolen~ borrowed from https://github.com/hashicorp/consul/commit/3e20bd25bdbec9b1b50fb5b1e0591314ec0c5fe2 😅

File: .github/workflows/build.yml
Comment (DanStough) [reply .github/workflows/build.yml:176]: Is it because ENT is on a different GO version? Is that intentional?

File: sdk/testutil/testlog.go
Comment (DanStough) [reply sdk/testutil/testlog.go:10]: I thought this was causing a linting error since the package is used as `hclog` but looks like I was wrong. Great catch!

File: tlsutil/config.go
Comment (DanStough) [reply tlsutil/config.go:467]: Yeah, that's a good call. I had a previous comment about how I needed to check something but wasn't familiar enough with `reflect`. I think I know how to handle it now.

File: agent/connect_ca_endpoint_test.go
Comment (DanStough) [reply agent/connect_ca_endpoint_test.go:294]: Originally this seemed like a pretty low value assertion like we were just testing the certPool add function, but I see now we're trying to validate the endpoint is returning the right number of cert. I found another way to do this.

File: tlsutil/config.go
Comment (DanStough) [reply tlsutil/config.go:467]: Actually this is the same code block RB pointed out above with my proposed soln. Will mark this resolved and we can move the discussion there.

File: connect/service_test.go
Comment (rboyer) [inline connect/service_test.go:294]: Maybe drop a TODO here for the future to reevaluate the kind of comparison we _should_ do here.

File: .github/workflows/build.yml
Comment (rboyer) [reply .github/workflows/build.yml:176]: `.circleci`, `.github`, `.release`, and `GNUMakefile` usually have collisions in general because of required divergences

**Review by rboyer — APPROVED** [review]

**Comment by hc-github-team-consul-core** [comment]
:cherries: If backport labels were added before merging, cherry-picking will start automatically.

To retroactively trigger a backport after merging, add backport labels and re-run https://circleci.com/gh/hashicorp/consul/642045.

**Comment by hc-github-team-consul-core** [comment]
:cherries::white_check_mark: Cherry pick of commit 95250e7915d07ba17217a7374cf96447a731142e onto `release/1.12.x` succeeded!

**Pull request merged**
